File: src/Concurrency/PrintFooBarAlternately.java
package Concurrency;

// https://leetcode.com/problems/print-foobar-alternately/
public class PrintFooBarAlternately {
    int n = 10;
    boolean fooPrinted = false;

    public synchronized void foo() throws InterruptedException {
        for (int i = 0; i < n; i++) {
            if (fooPrinted) wait();
            System.out.print("foo");
            fooPrinted = true;
            notifyAll();
        }
    }

    public synchronized void bar() throws InterruptedException {
        for (int i = 0; i < n; i++) {
            if (!fooPrinted) wait();
            System.out.println("bar");
            fooPrinted = false;
            notifyAll();
        }
    }

    public static void main(String[] args) {
        PrintFooBarAlternately p = new PrintFooBarAlternately();

        Thread t1 = new Thread(() -> {
            try {
                p.foo();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });

        Thread t2 = new Thread(() -> {
            try {
                p.bar();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });

        t1.start();
        t2.start();

    }
}

File: py/concurrency/PrintFooBarAlternately.py
# https://leetcode.com/problems/print-foobar-alternately/description/?envType=problem-list-v2&envId=concurrency

import threading
from typing import Callable


class FooBar:
    def __init__(self, n):
        self.n = n
        self.lockFoo = threading.Lock()
        self.lockBar = threading.Lock()
        self.lockBar.acquire()

    def foo(self, printFoo: "Callable[[], None]") -> None:
        for i in range(self.n):
            with self.lockFoo:
                printFoo()
                self.lockBar.release()
            self.lockFoo.acquire()

    def bar(self, printBar: "Callable[[], None]") -> None:
        for i in range(self.n):
            with self.lockBar:
                printBar()
                self.lockFoo.release()
            self.lockBar.acquire()


fb = FooBar(5)


def printFoo():
    print("foo")


def printBar():
    print("bar")


threadA = threading.Thread(target=fb.foo, args=(printFoo,), name="threadA")
threadB = threading.Thread(target=fb.bar, args=(printBar,), name="threadB")
threadA.start()
threadB.start()

threadA.join()
threadB.join()
